optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -15,180 +15,207 @@
-
-        # (batch_size, n_heads, seq_length, key_length)
-        attn_weights = nn.functional.softmax(attention_scores.float(), dim=-1).to(value_states.dtype)
-        attn_weights = nn.functional.dropout(attn_weights, p=self.attn_dropout_p, training=self.training)
-
-        attn_output = torch.matmul(attn_weights, value_states)
-
-    attn_output = attn_output.permute(0, 2, 1, 3).contiguous().view(batch_size, seq_length, -1)
-    attn_output = self.out_proj(attn_output)
-
-    return attn_output, attn_weights, past_key_value
-
-
-def gaudi_mpt_block_forward(
-    self,
-    hidden_states: torch.Tensor,
-    position_bias: torch.Tensor,
-    attention_mask: torch.Tensor,
-    layer_past: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
-    use_cache: bool = False,
-    output_attentions: bool = False,
-    token_idx: Optional[torch.Tensor] = None,
-    use_flash_attention: Optional[bool] = False,
-    flash_attention_recompute: Optional[bool] = False,
-):
-    """
-    Copied from MptBlock.forward: https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/models/mpt/modeling_mpt.py
-    The only differences are:
-    - add new args token_idx
-    - add new args use_flash_attention
-    - add new arg flash_attention_recompute
-    """
-    # hidden_states: [batch_size, seq_length, hidden_size]
-    # Layer norm at the beginning of the transformer layer.
-    layernorm_output = self.norm_1(hidden_states)
-
-    residual = hidden_states
-
-    # Self attention.
-    attn_outputs, attn_weights, past_key_value = self.attn(
-        layernorm_output,
-        position_bias=position_bias,
-        attention_mask=attention_mask,
-        past_key_value=layer_past,
-        token_idx=token_idx,
-        use_flash_attention=use_flash_attention,
-        flash_attention_recompute=flash_attention_recompute,
-    )
-
-    hidden_states = self.resid_attn_dropout(attn_outputs) + residual
-
-    layernorm_output = self.norm_2(hidden_states)
-
-    # Get residual
-    residual = hidden_states
-
-    # MLP.
-    output = self.ffn(layernorm_output, residual)
-    outputs = (output,)
-
-    if use_cache:
-        outputs += (past_key_value,)
-
-    if output_attentions:
-        outputs += (attn_weights,)
-
-    return outputs  # hidden_states, present, attentions
+            if len(position_bias.shape) != 3:
+                raise ValueError(f"Expecting position_bias shape to be 3 dimensions, got {len(position_bias.shape)}")
+            key_length = key_states.shape[-2]
+
+            position_bias_query_index = max(0, position_bias.size(1) - query_length)
+            position_bias_key_index = max(0, position_bias.size(2) - key_length)
+
+            position_bias = position_bias[:, position_bias_query_index:, position_bias_key_index:]
+
+        if use_flash_attention and FusedSDPA:
+            import habana_frameworks.torch.hpu as ht
+
+            with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
```

Comment:
Can you also check other model if this enable_recompute should be set based on the fp8/bf16 and q_len.

Reply:
If set enable_recompute same way of other models, then FP8 throughput drop to half. In the trace, softmax_stage1_fwd_f32 appears and spending a lot of time, There is not this process at enable_recompute =  False case.

Reply:
what's your fp8 command with flash_attention?

Reply:
Did you test any case which enable causal_mask and enable_compute? 
How about longer prompt? Usually causal_mask shows better perf for long prompt

Reply:
> what's your fp8 command with flash_attention?
The command line is 
```
QUANT_CONFIG=./quantization_config/maxabs_quant.json \
python run_generation.py \
--model_name_or_path mosaicml/mpt-7b \
--use_hpu_graphs \
--use_kv_cache \
--limit_hpu_graphs \
--max_input_tokens 128 \
--max_new_tokens 128 \
--batch_size 128 \
--bf16 \
--use_flash_attention
```

Reply:
> Did you test any case which enable causal_mask and enable_compute? How about longer prompt? Usually causal_mask shows better perf for long prompt

No, I haven't test it these cases.